### PR TITLE
MatMulScaleFusion - handling scale input

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: this is to handle an edge case reported here:
https://github.com/microsoft/onnxruntime/issues/10950

The proposed solution is not to fuse mul and matmul when mul's input is scalar.

User should follow the same code here to run Inference engine with proper input:
https://github.com/microsoft/onnxruntime/issues/1166

sample code:
x1 = np.array((1.0)).astype(np.float32)
session = ort.InferenceSession("/bert_ort/ettao/dri-test/matnulfuse/1.onnx", providers=['CUDAExecutionProvider', 'CPUExecutionProvider'])
res = session.run(None,{"i1":x1})

### CODE REVIEW

**Comment by pengwa** [comment]
I did not get why FusedMatMul fail to handle this case, could you pls elaborate on this?

**Comment by ytaous** [comment]
> I did not get why FusedMatMul fail to handle this case, could you pls elaborate on this?

In the original code, after initial transformation for 1.onnx, the FusedMatMul would have two inputs. One of the inputs, i1, would fail the shape check in contrib_def.cc since its shape has 0 dim. 

  if (shape0_raw.dim_size() == 0 || shape1_raw.dim_size() == 0) {
    fail_shape_inference("Input tensors of wrong rank (0).");
  }

**Comment by ytaous** [comment]
> Instead of adding an unsqueeze node can we simply not fuse the 2 nodes when the input is scalar?

That's my initial thought, let me test it out and see if it goes thru, thx

**Comment by pengwa** [comment]
> > I did not get why FusedMatMul fail to handle this case, could you pls elaborate on this?
> 
> In the original code, after initial transformation for 1.onnx, the FusedMatMul would have two inputs. One of the inputs, i1, would fail the shape check in contrib_def.cc since its shape has 0 dim.
> 
> if (shape0_raw.dim_size() == 0 || shape1_raw.dim_size() == 0) { fail_shape_inference("Input tensors of wrong rank (0)."); }

Is it better we let FusedMatMul support the scalar input (shape inferencing and kernel)?

**Comment by ytaous** [comment]
> > > I did not get why FusedMatMul fail to handle this case, could you pls elaborate on this?
> > 
> > 
> > In the original code, after initial transformation for 1.onnx, the FusedMatMul would have two inputs. One of the inputs, i1, would fail the shape check in contrib_def.cc since its shape has 0 dim.
> > if (shape0_raw.dim_size() == 0 || shape1_raw.dim_size() == 0) { fail_shape_inference("Input tensors of wrong rank (0)."); }
> 
> Is it better we let FusedMatMul support the scalar input (shape inferencing and kernel)? I guess we just need add some check for rank=0, and adapt it to rank=1, subsequent logic should remain unchanged.

So far, this is edge case. If we encounter 1p/3p models in future that actually have this case, we can then enhance it further.

**Comment by pengwa** [comment]
> FusedMatMul

You are right, rank=0 should be invalid for FusedMatMul so far. The issue user complains is due to the incorrect fusion.

File: onnxruntime/core/optimizer/matmul_scale_fusion.cc [hunk 1]
```diff
@@ -131,6 +131,11 @@ std::vector<ScaleMergeInfo> GetInputNodeMerges(
     ORT_ENFORCE(input_node.InputDefs().size() == 2 && scale_and_index->second < 2);
     const int to_scale_index = 1 - scale_and_index->second;
 
+    // check if the non-scale input is graph input and scalar 
+    const auto non_scale_node_arg = input_node.InputDefs()[to_scale_index];
+    const auto* shape = non_scale_node_arg->Shape();
```
Comment (pengwa) [inline onnxruntime/core/optimizer/matmul_scale_fusion.cc:136]: > shape = non_scale_node_arg->Shape(); [](http://example.com/codeflow?start=15&length=37)

need check shape is nullptr or not.

File: onnxruntime/core/optimizer/matmul_scale_fusion.cc [hunk 2]
```diff
@@ -131,6 +131,11 @@ std::vector<ScaleMergeInfo> GetInputNodeMerges(
...
+    if (graph_utils::IsGraphInput(graph, non_scale_node_arg) && shape->dim_size() == 0) continue;
```
Comment (pengwa) [inline onnxruntime/core/optimizer/matmul_scale_fusion.cc:137]: >IsGraphInput [](http://example.com/codeflow?start=21&length=12)

I think we don't need check it is a graph input, any reason for doing this?

Comment (ytaous) [reply onnxruntime/core/optimizer/matmul_scale_fusion.cc:137]: right, we don't need it

File: onnxruntime/core/optimizer/matmul_scale_fusion.cc [hunk 3]
```diff
@@ -131,6 +131,11 @@ std::vector<ScaleMergeInfo> GetInputNodeMerges(
     ORT_ENFORCE(input_node.InputDefs().size() == 2 && scale_and_index->second < 2);
     const int to_scale_index = 1 - scale_and_index->second;
 
+    // check if the non-scale input is graph input and scalar
```
Comment (pengwa) [inline onnxruntime/core/optimizer/matmul_scale_fusion.cc:134]: >graph input [](http://example.com/codeflow?start=39&length=11)

pls update the comment accordingly.

Comment (ytaous) [reply onnxruntime/core/optimizer/matmul_scale_fusion.cc:134]: good call

**Review by pengwa — APPROVED** [review]
:shipit:

**Pull request merged**